modules/dnn/src/layers/gather_elements_layer.cpp

```diff
@@ -0,0 +1,148 @@
+
+        const Mat& data = inputs[0];
+        const Mat& indices = inputs[1];
+        Mat& out = outputs[0];
+
+        typeDispatch(outputs[0].type(), data, indices, out);
+    }
+
+    template<typename T, typename Functor>
+    void forward_impl(const Functor& rd, const Mat& data, const Mat& indices,  Mat& out)
+    {
+
+        const int ndims = data.dims;
+        CV_Assert(ndims >= 1);
+        CV_Assert(axis >= -1 * ndims && axis <= (ndims - 1));
+
+        const int indices_ndims = indices.dims;
+        CV_CheckEQ(data.dims, indices.dims, "GatherElements: input and indices have to be of same rank.");
+
+        const int* shape = data.size.p;
+        const size_t* step = data.step.p;
+
+        const int* ind_shape = indices.size.p;
+        const size_t* ind_step = indices.step.p;
+
+        size_t inp_offset = 0;
+        size_t ind_offset = 0;
+        const T* p_index = indices.ptr<const T>();
+        const T* p_data = data.ptr<const T>();
+        T* p_out = out.ptr<T>();
+
+        size_t total = indices.total();
+
+        int j, offset_at_idx, index;
+        size_t t, idx;
+        for (size_t i = 0; i < total; i++)
+        {
+            t = i;
+            inp_offset = 0;
+            ind_offset = 0;
+            int offset_at_axis = 0;
+            for (j = ndims - 1; j >= 0; j--)
+            {
+                idx = t / ind_shape[j];
+                offset_at_idx = (int)(t - idx * ind_shape[j]);
+                ind_offset += offset_at_idx * ind_step[j];
+                inp_offset += offset_at_idx * step[j];
+                t = idx;
+                if (j == axis)
+                {
+                    offset_at_axis = offset_at_idx * step[j];
+                }
+            }
+            ind_offset /= sizeof(T);
+
+            // get index and overwrite current indices
+            const T* tmp_p_index = p_index + ind_offset;
+            index = (int)(*tmp_p_index);
+            CV_Assert(index < shape[axis] && index > -shape[axis]);
+
+            inp_offset = inp_offset - offset_at_axis + ((index + shape[axis]) % shape[axis]) * step[axis];
+            inp_offset /= sizeof(T);
+
+            const T* tmp_p_data = p_data + inp_offset;
+            T* tmp_p_out = p_out + ind_offset;
+            *tmp_p_out = rd(*tmp_p_out, *tmp_p_data);
+        }
+    }
+
+    template<typename... Args>
+    inline void typeDispatch(const int type, Args&&... args)
+    {
+        switch (type)
+        {
+            case CV_8U:
+                reductionDispatch<uint8_t>(std::forward<Args>(args)...);
+                break;
+            case CV_32S:
+                reductionDispatch<int32_t>(std::forward<Args>(args)...);
+                break;
+            case CV_32F:
+                reductionDispatch<float>(std::forward<Args>(args)...);
+                break;
+            default:
+                CV_Error(cv::Error::BadDepth, "Unsupported type.");
+        };
+    }
+
+    template<typename T, typename... Args>
```

Comment:
There is no reduction operations in `GatherElements`. So we most likely do not need this dispatcher.

Reply:
removed it in the second commit

Reply:
I reverted to the first commit since it was working correctly and passed the tests. Essentially, I edited the dispatcher from the original implementation of scatter, and this dispatcher essentially just calls forward_impl<T>(). I decided to keep the dispatchers just to account for the data type of the output and since they were in harmony with the rest of the implementation in terms of the use of the template.

Reply:
Extra calls consume extra time. We should try to save time, especially when it's inference. You can try something like this:

```cpp
template<typename T>
void forward_impl(const Mat& data, const Mat& indices,  Mat& out) {
// ...
for (size_t i = 0; i < total; i++) {
    *tmp_p_out = *tmp_p_data;
}
}

template<typename... Args>
inline void typeDispatch(const int type, Args&&... args)
{
    switch (type)
    {
        case CV_8U:
            forward_impl<uint8_t>(std::forward<Args>(args)...);
            break;
        case CV_32S:
            forward_impl<int32_t>(std::forward<Args>(args)...);
            break;
        case CV_32F:
            forward_impl<float>(std::forward<Args>(args)...);
            break;
        default:
            CV_Error(cv::Error::BadDepth, "DNN/GatherElements: Unsupported type.");
    };
}
```

Reply:
Done